### PR TITLE
Quit command to exit dotnet-interactive

### PR DESCRIPTION
**Pull request by colombod**
The Pr provides protocol extensions for http and stdio mode so that submitting a quit command allows for exiting the process with a 0 code.
Can terminate by
- CTRL+C 
- posting a command `{ commandType: " Quit" ,  command: {}}`

Relates to #385

### CODE REVIEW

File: src/dotnet-interactive/CommandLine/CommandLineParser.cs [hunk 1]
```diff
@@ -290,6 +293,7 @@ Command StdIO()
                 command.Handler = CommandHandler.Create<StartupOptions, StdIOOptions, IConsole, InvocationContext>(
                     (startupOptions, options, console, context) =>
                     {
+                        ExtendProtocol();
```
Comment (jonsequitur) [inline src/dotnet-interactive/CommandLine/CommandLineParser.cs:296]: Can we give this a name that explains the why rather than the what?

File: src/dotnet-interactive/Commands/Quit.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.DotNet.Interactive.Commands;
+
+namespace Microsoft.DotNet.Interactive.App.Commands
+{
+    public class Quit : KernelCommandBase
+    {
+        public Quit(Action processTermination = null, string targetKernelName = null): base(targetKernelName)
```
Comment (jonsequitur) [inline src/dotnet-interactive/Commands/Quit.cs:13]:
```suggestion
        public Quit(Action onQuit = null, string targetKernelName = null): base(targetKernelName)
```

It's good to avoid using two different words that mean effectively the same thing.

File: src/dotnet-interactive.Tests/CommandLine/CommandLineParserTests.cs [hunk 3]
```diff
@@ -373,6 +387,17 @@ public void stdio_command_honors_default_kernel_option()
...
-       
+        [Fact]
+        public async Task stdio_command_extends_the_protocol_with_quit_command()
+        {
+            await _parser.InvokeAsync("stdio");
+
+            var envelope = KernelCommandEnvelope.Deserialize(@"{ commandType: ""Quit"", command : { } }");
+            
+            envelope.Command.Should()
+                .NotBeNull()
```
Comment (jonsequitur) [inline src/dotnet-interactive.Tests/CommandLine/CommandLineParserTests.cs:398]: I think `NotBeNull` is redundant here.

File: src/dotnet-interactive/CommandLine/CommandLineParser.cs (hunk 1 above)
Comment (colombod) [reply src/dotnet-interactive/CommandLine/CommandLineParser.cs:316]: ```UseQuitCommand()``` ?

File: src/Microsoft.DotNet.Interactive/InputTextStream.cs [hunk 4]
```diff
@@ -46,7 +51,7 @@ private void EnsureStarted()
 
             Task.Run(async () =>
             {
-                while (!_complete)
+                while (!_complete && !_cancellationSource.IsCancellationRequested)
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive/InputTextStream.cs:54]: Please don't do this to my CPU.

File: src/Microsoft.DotNet.Interactive/InputTextStream.cs [hunk 5]
```diff
@@ -58,13 +63,15 @@ private void EnsureStarted()
                         _channel.OnNext(line);
                     }
                 }
-            });
+            }, _cancellationSource.Token);
         }
 
         public void Dispose()
         {
+            _channel.OnNext(string.Empty);
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive/InputTextStream.cs:71]: This looks odd. Why is this needed?

File: src/dotnet-interactive/Commands/Quit.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+namespace Microsoft.DotNet.Interactive.App.Commands
+{
+    public class Quit : KernelCommandBase
+    {
+
+        public Quit(Action onQuit = null, string targetKernelName = null): base(targetKernelName)
+        {
+            onQuit ??= DefaultOnQuit ??(() =>
+            {
+                Environment.Exit(0);
```
Comment (jonsequitur) [inline src/dotnet-interactive/Commands/Quit.cs:18]: The second fallback is in case `DefaultOnQuit` is `null`. Why allow `DefaultOnQuit` to be `null`?

File: src/dotnet-interactive/CommandLine/CommandLineParser.cs [hunk 7]
```diff
@@ -304,15 +328,36 @@ Command StdIO()
...
 
-                        return startStdIO(
-                            startupOptions,
-                            CreateKernel(options.DefaultKernel, new BrowserFrontendEnvironment(), startupOptions),
-                            console);
+                        {
+                            var kernel = CreateKernel(options.DefaultKernel, new BrowserFrontendEnvironment(),
+                                startupOptions);
+                            
+                            Quit.DefaultOnQuit = () =>
```
Comment (jonsequitur) [inline src/dotnet-interactive/CommandLine/CommandLineParser.cs:345]: Any reason not to hardcode the default quit behavior?

File: src/Microsoft.DotNet.Interactive/InputTextStream.cs (hunk 5 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive/InputTextStream.cs:71]: Completing that steam without any value causes LastAsync to throw as there are now value for the async subject

File: src/dotnet-interactive/CommandLine/CommandLineParser.cs (hunk 7 above)
Comment (colombod) [reply src/dotnet-interactive/CommandLine/CommandLineParser.cs:345]: that is the only time i am overriding it

File: src/dotnet-interactive/Commands/Quit.cs (hunk 6 above)
Comment (colombod) [reply src/dotnet-interactive/Commands/Quit.cs:18]: makes sense, setting it to null thrown

File: src/dotnet-interactive/CommandLine/CommandLineParser.cs (hunk 1 above)
Comment (colombod) [reply src/dotnet-interactive/CommandLine/CommandLineParser.cs]: Done

File: src/dotnet-interactive/Commands/Quit.cs (hunk 2 above)
Comment (colombod) [reply src/dotnet-interactive/Commands/Quit.cs]: Done

File: src/dotnet-interactive.Tests/CommandLine/CommandLineParserTests.cs (hunk 3 above)
Comment (colombod) [reply src/dotnet-interactive.Tests/CommandLine/CommandLineParserTests.cs]: done

File: src/Microsoft.DotNet.Interactive/InputTextStream.cs (hunk 4 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive/InputTextStream.cs]: it was fine on my Quantum processor but can see the point

File: src/Microsoft.DotNet.Interactive.Jupyter/JupyterRequestContextHandler.cs [hunk 8]
```diff
@@ -16,6 +16,7 @@ public class JupyterRequestContextHandler : ICommandHandler<JupyterRequestContex
         private readonly CompleteRequestHandler _completeHandler;
         private readonly InterruptRequestHandler _interruptHandler;
         private readonly IsCompleteRequestHandler _isCompleteHandler;
+        private readonly ShutdownRequestHandler _shutDownHandler;
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Jupyter/JupyterRequestContextHandler.cs:19]:
```suggestion
        private readonly ShutdownRequestHandler _shutdownHandler;
```

File: src/Microsoft.DotNet.Interactive.Jupyter/JupyterRequestContextHandler.cs [hunk 9]
```diff
@@ -29,6 +30,7 @@ public JupyterRequestContextHandler(IKernel kernel)
             _completeHandler = new CompleteRequestHandler(kernel, scheduler);
             _interruptHandler = new InterruptRequestHandler(kernel, scheduler);
             _isCompleteHandler = new IsCompleteRequestHandler(kernel, scheduler);
+            _shutDownHandler = new ShutdownRequestHandler(kernel, scheduler);
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Jupyter/JupyterRequestContextHandler.cs:33]:
```suggestion
            _shutdownHandler = new ShutdownRequestHandler(kernel, scheduler);
```

File: src/Microsoft.DotNet.Interactive.Jupyter/JupyterRequestContextHandler.cs [hunk 10]
```diff
@@ -48,6 +50,9 @@ public JupyterRequestContextHandler(IKernel kernel)
                 case IsCompleteRequest _:
                     await _isCompleteHandler.Handle(delivery.Command);
                     break;
+                case ShutdownRequest _:
+                    await _shutDownHandler.Handle(delivery.Command);
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Jupyter/JupyterRequestContextHandler.cs:54]:
```suggestion
                    await _shutdownHandler.Handle(delivery.Command);
```

File: src/Microsoft.DotNet.Interactive.Jupyter/ShutdownRequestHandler.cs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        public async Task Handle(JupyterRequestContext context)
+        {
+            // reply 
+            var shutdownReplyPayload = new ShutdownReply();
+
+            // send to server
+            context.JupyterMessageSender.Send(shutdownReplyPayload);
+            
+            // pause and then exit
+            await Task.Delay(500);
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Jupyter/ShutdownRequestHandler.cs:34]: Is the pause needed?

File: src/Microsoft.DotNet.Interactive/Server/KernelCommandEnvelope.cs [hunk 12]
```diff
@@ -17,11 +17,40 @@ public abstract class KernelCommandEnvelope : IKernelCommandEnvelope
...
+        public string Token => _command.GetToken();
+
+        IKernelCommand IKernelCommandEnvelope.Command => _command;
+
+        public static void RegisterCommandType<T>(string commandTypeName) where T : class, IKernelCommand
+        {
+            var commandEnvelopeType = typeof(KernelCommandEnvelope<T>);
+            var commandType = typeof(T);
+
+            _envelopeTypesByCommandTypeName[commandTypeName] = commandEnvelopeType;
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive/Server/KernelCommandEnvelope.cs:50]: I wonder if we want to use `Add` so that it throws if the command type is already added.

File: src/Microsoft.DotNet.Interactive/Server/StandardIOKernelServer.cs [hunk 13]
```diff
@@ -39,7 +39,8 @@ public class StandardIOKernelServer : IDisposable
                 {
                     await DeserializeAndSendCommand(line);
                 }),
-                _kernel.KernelEvents.Subscribe(WriteEventToOutput)
+                _kernel.KernelEvents.Subscribe(WriteEventToOutput),
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive/Server/StandardIOKernelServer.cs:42]: Should we be tracking the subscription for disposal?

File: src/dotnet-interactive/CommandLine/CommandLineParser.cs [hunk 14]
```diff
@@ -353,6 +392,12 @@ static HttpPortRange ParsePortRangeOption(ArgumentResult result)
             }
         }
 
+        private static void UseQuitCommand(IDisposable disposeOnQuit)
+        {
+            Quit.DisposeOnQuit = disposeOnQuit;
+            KernelCommandEnvelope.RegisterCommandType<Quit>(nameof(Quit));
```
Comment (jonsequitur) [inline src/dotnet-interactive/CommandLine/CommandLineParser.cs:398]: Doing this without a static would be preferable.

It seems like adding the command to the kernel, including the implementation, could avoid this, e.g.

```csharp
kernel.UseCommand<Quit>();
```

File: src/Microsoft.DotNet.Interactive.Jupyter/JupyterRequestContextHandler.cs (hunk 8 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Jupyter/JupyterRequestContextHandler.cs:19]: done

File: src/Microsoft.DotNet.Interactive.Jupyter/JupyterRequestContextHandler.cs (hunk 9 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Jupyter/JupyterRequestContextHandler.cs:33]: done

File: src/Microsoft.DotNet.Interactive.Jupyter/JupyterRequestContextHandler.cs (hunk 10 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Jupyter/JupyterRequestContextHandler.cs:54]: done

File: src/Microsoft.DotNet.Interactive.Jupyter/ShutdownRequestHandler.cs (hunk 11 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Jupyter/ShutdownRequestHandler.cs:34]: no

File: src/Microsoft.DotNet.Interactive/Server/KernelCommandEnvelope.cs (hunk 12 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive/Server/KernelCommandEnvelope.cs:50]: I think is a good point

File: src/Microsoft.DotNet.Interactive/Server/StandardIOKernelServer.cs (hunk 13 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive/Server/StandardIOKernelServer.cs:42]: we do, that code  is inside the CompositeDisposabe contructor

File: src/dotnet-interactive/CommandLine/CommandLineParser.cs (hunk 14 above)
Comment (colombod) [reply src/dotnet-interactive/CommandLine/CommandLineParser.cs:398]: liked it and did something along that line

File: src/Microsoft.DotNet.Interactive/Server/StandardIOKernelServer.cs (hunk 13 above)
Comment (jonsequitur) [reply src/Microsoft.DotNet.Interactive/Server/StandardIOKernelServer.cs:42]: Subscribe returns a disposable and we're not keeping a reference to it.
